rust/src/enhancers/grammar.rs

```diff
@@ -1,179 +1,350 @@
+    };
+
+    let (name, after_name) = ident(before_name)
+        .with_context(|| format!("at `{before_name}`: failed to parse matcher name"))?;
+
+    let before_arg = expect(after_name, ":")?;
+
+    let (arg, rest) = argument(before_arg)
+        .with_context(|| format!("at `{before_arg}`: failed to parse matcher argument"))?;
+
+    let m = Matcher::new(negated, name, &arg, frame_offset, regex_cache)?;
+    Ok((m, rest))
+}
+
+/// Parses the caller matcher in a rule and returns it together with the rest of the input.
+///
+/// A caller matcher is defined as `[ <matcher> ] |`.
+/// NB: This function assumes that the leading `[` has already been consumed!
+fn caller_matcher<'a>(
+    input: &'a str,
+    regex_cache: &mut RegexCache,
+) -> anyhow::Result<(Matcher, &'a str)> {
+    let (matcher, rest) = matcher(input, FrameOffset::Caller, regex_cache)?;
+
+    let rest = rest.trim_start();
+    let rest = expect(rest, "]")?;
+
+    let rest = rest.trim_start();
+    let rest = expect(rest, "|")?;
+
+    Ok((matcher, rest))
+}
+
+/// Parses the callee matcher in a rule and returns it together with the rest of the input.
+///
+/// A callee matcher is defined as `| [ <matcher> ] `.
+/// NB: This function assumes that the leading `|` has already been consumed!
+fn callee_matcher<'a>(
+    input: &'a str,
+    regex_cache: &mut RegexCache,
+) -> anyhow::Result<(Matcher, &'a str)> {
+    let rest = input.trim_start();
+    let rest = expect(rest, "[")?;
+
+    let (matcher, rest) = matcher(rest, FrameOffset::Callee, regex_cache)?;
+
+    let rest = rest.trim_start();
+    let rest = expect(rest, "]")?;
+
+    Ok((matcher, rest))
+}
+
+/// Parses a sequence of [`Matchers`](Matcher) and returns it
+/// together with the rest of the input.
+///
+/// The sequence must contain at least one matcher.
+fn matchers<'a>(
+    input: &'a str,
+    regex_cache: &mut RegexCache,
+) -> anyhow::Result<(Vec<Matcher>, &'a str)> {
+    let mut input = input.trim_start();
+
+    let mut result = Vec::new();
+
+    // A `[` at the start means we have a caller matcher
+    if let Some(rest) = input.strip_prefix('[') {
+        let (caller_matcher, rest) = caller_matcher(rest, regex_cache)
+            .with_context(|| format!("at `{input}`: failed to parse caller matcher"))?;
+
+        result.push(caller_matcher);
+
+        input = rest.trim_start()
+    }
+
+    // Keep track of whether we've parsed at least one matcher
+    let mut parsed = false;
+
+    // Possible prefixes of a matcher definition.
+    // Matchers always start with one of these,
+    // and actions never do. This means that if
+    // the rest of the input starts with one these,
+    // there is another matcher to parse, and if it doesn't,
+    // there isn't.
+    const MATCHER_LOOKAHEAD: [&str; 11] = [
+        "!",
+        "a",
+        "category:",
+        "e",
+        "f",
+        "me",
+        "mo",
+        "p",
+        "s",
+        "t",
+        "va",
+    ];
```

Comment:
This can be static and declared outside of this scope

Reply:
I originally had it outside, I'll move it back.

Reply:
Static is more performant afaik